TracingId according to w3c standard

msallin opened the pull request:
Reopen this draft PR, as certain things changed according to tracing in .NET Core 3.0 and because this contribution is part of work for a Swiss Post project, hence the work will be done under swisspush.

msallin commented on tests/Silverback.Integration.Kafka.TestConsumer/Program.cs:
```diff
@@ -45,7 +54,8 @@ private static void Connect()
                     AutoOffsetReset = Confluent.Kafka.AutoOffsetReset.Earliest
-                }
+                },
+                Serializer = new JsonMessageSerializer<TestMessage>()
```
Shouldn't be necessary.

msallin replied:
It is necessary, because the FullName of the type is used and not only the Namespace.

msallin commented:
- Move to own class but don't use behavior as this gets called to late
- Extend examples
- Write Unit Tests

BEagle1984 commented on src/Silverback.Integration/Diagnostics/DiagnosticsConsumer.cs:
```diff
@@ -0,0 +1,81 @@
+
+namespace Silverback.Diagnostics
+{
+    public abstract class DiagnosticsConsumer : Consumer
```
It works properly with the intermediate base class `DiagnosticsConsumer` and `DiagnosticsProducer` but I don't know if I like it. I see this as a side aspect and I would have just used an helper from within the regular `Consumer`.

Also, when other stuff get added, it will be confusing for whoever implements a extra consumer or producer to choose which is the proper class to extend and doesn't give the freedom to choose exactly which aspect to inherit.

I think we should move this logic into something pluggable...but it can be done later on.

What I imagine (completely made up right now):

```c#
public abstract class Consumer : ...
{
    public readonly IEnumerable<IConsumerBehavior> _behaviors;
    public Consumer(IEnumerable<IConsumerBehavior> behaviors) => _behaviors = behaviors;
    
        protected Task HandleMessage(byte[] message, IEnumerable<MessageHeader> headers, IOffset offset)
        {
            _behaviors.ForEach(b => b.Consuming(message, headers, offset));

            try
            {
                //...actual consume logic...

                _behaviors.ForEach(b => b.Consumed(message, headers, offset));
            }
            catch (Exception ex)
            {
                _behaviors.ForEach(b => b.Failed(message, headers, offset));
            }
        }
}
```

So that trancing, logging, activity and all other aspects can be moved into an implementation of `IConsumerBehavior` and easily activated/deactivated in the startup.

The same can of course easily be adapted into an `IProducerBehavior` (with the difference that it can also modify the actual message being sent).

Anyhow, I suggest to make it work like this and refactor it later. If you don't mind, I can provide the basic implementation/prototype for these "behaviors" in a branch, maybe tomorrow while traveling to Bern.

BEagle1984 replied:
Note: ...discussion is going on, this is not up-to-date anymore

BEagle1984 replied:
The `IConsumerBehavior` and `IProducerBehavior` could be implemented following the same logic of the `IBehavior`.

https://github.com/BEagle1984/silverback/blob/75078c8aec459bae32a027fc069eef8099db771c/src/Silverback.Core/Messaging/Publishing/Publisher.cs#L85

https://github.com/BEagle1984/silverback/blob/75078c8aec459bae32a027fc069eef8099db771c/src/Silverback.Core/Messaging/Publishing/IBehavior.cs#L13

BEagle1984 commented on src/Silverback.Integration/Silverback.Integration.csproj:
```diff
@@ -36,5 +36,7 @@ This package contains the message broker and connectors abstractions.</Descripti
     <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
     <PackageReference Include="Silverback.Core" Version="$(BaseVersion)" />
     <PackageReference Include="System.Memory" Version="4.5.2" />
+    <PackageReference Include="System.Memory" Version="4.5.3" />
```
Something is still broken for me. It doens't compile on azure devops nor on my mac.

msallin replied:
Jup, will fix this after we got the implementation details correct.

msallin commented on src/Silverback.Integration/Diagnostics/BaggageConverter.cs:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    public class DiagnosticsConstants
```
Done

msallin commented on src/Silverback.Integration/Diagnostics/BaggageConverter.cs:
```diff
@@ -0,0 +1,80 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Silverback.Diagnostics
```
Will run a R# cleanup

msallin commented on src/Silverback.Integration/Messaging/Messages/OutboundMessage.cs:
```diff
@@ -7,7 +7,7 @@
 namespace Silverback.Messaging.Messages
 {
-    internal class OutboundMessage : IOutboundMessageInternal
+    public class OutboundMessage : IOutboundMessageInternal
```
Because it is exposed on a method which is protected and the abstract class is public. But I try to find a way to prevent this.

msallin commented:
As we are unsure how to continue with the DiagnosticsConsumer & DiagnosticsProducer, I moved all logic out of them. Then, I added unit tests for the new functionality.
The following classes aren't yet covered, as they might get deleted.
- DiagnosticsConsumer 
- DiagnosticsProducer 
- ActivityScope

Furthermore, there is still the visibility of "OutboundMessage" which will also change as part of the refactoring.

As discussed, there is a problem with the dependency to system.memory.

In addition, I didn't wrote any integration tests, as I didn't want to include the classes to be refactored (DiagnosticsConsumer & DiagnosticsProducer)

msallin commented on src/Silverback.Integration/Diagnostics/ActivityExtensions.cs:
```diff
@@ -0,0 +1,64 @@
+
+        public static void FillActivity(this Activity activity, IBrokerMessage brokerMessage)
+        {
+            brokerMessage.Headers.Add(new MessageHeader(DiagnosticsConstants.TraceIdHeaderKey, activity.Id));
```
Check activity.Id for null and throw an InvalidOperationException if its null. The activityId is null, if there is no started activity. As it makes no sense to call "FillActivity" in this case, this is a behavior which should never happen. Thus, the exception notifies the dev that he uses the code the wrong way.
